modelzoo/DIN/train.py

```diff
@@ -1,387 +1,739 @@
+
+        return uid_emb, item_emb, his_item_emb, sequence_length
+
+    def top_fc_layer(self, inputs):
+        bn1 = tf.layers.batch_normalization(inputs=inputs, name='bn1')
+        dnn1 = tf.layers.dense(bn1, 200, activation=None, name='dnn1')
+        dnn1 = self.dice(dnn1, name='dice_1')
+
+        dnn2 = tf.layers.dense(dnn1, 80, activation=None, name='dnn2')
+        dnn2 = self.dice(dnn2, name='dice_2')
+        dnn3 = tf.layers.dense(dnn2, 2, activation=None, name='dnn3')
+        logits = tf.layers.dense(dnn3, 1, activation=None, name='logits')
+        add_layer_summary(dnn1, 'dnn1')
+        add_layer_summary(dnn2, 'dnn2')
+        add_layer_summary(dnn3, 'dnn3')
+        return logits
+
+    def prediction(self):
+        # input layer to get embedding of features
+        with tf.variable_scope('input_layer',
+                               partitioner=self.input_layer_partitioner,
+                               reuse=tf.AUTO_REUSE):
+            uid_emb, item_emb, his_item_emb, sequence_length = self.embedding_input(
+            )
+            if self.bf16:
+                uid_emb = tf.cast(uid_emb, tf.bfloat16)
+                item_emb = tf.cast(item_emb, tf.bfloat16)
+                his_item_emb = tf.cast(his_item_emb, tf.bfloat16)
+
+            item_his_eb_sum = tf.reduce_sum(his_item_emb, 1)
+            mask = tf.sequence_mask(sequence_length)
+
+        # Attention layer
+        attention_scope = tf.variable_scope('attention_layer')
+        with attention_scope.keep_weights(dtype=tf.float32) if self.bf16 \
+            else attention_scope:
+            attention_output = self.attention(item_emb, his_item_emb,
+                                              self.attention_size, mask)
+            att_fea = tf.reduce_sum(attention_output, 1)
+            tf.summary.histogram('att_fea', att_fea)
+
+        top_input = tf.concat([
+            uid_emb, item_emb, item_his_eb_sum, item_emb * item_his_eb_sum,
+            att_fea
+        ], 1)
+        # Top MLP layer
+        top_mlp_scope = tf.variable_scope(
+            'top_mlp_layer',
+            partitioner=self.dense_layer_partitioner,
+            reuse=tf.AUTO_REUSE,
+        )
+        with top_mlp_scope.keep_weights(dtype=tf.float32) if self.bf16 \
+            else top_mlp_scope:
+            self.logits = self.top_fc_layer(top_input)
+        if self.bf16:
+            self.logits = tf.cast(self.logits, dtype=tf.float32)
+
+        predict = tf.math.sigmoid(self.logits)
+        return predict
+
+    def optimizer(self):
+        self.logits = tf.squeeze(self.logits)
+        loss = tf.losses.sigmoid_cross_entropy(
+            self.label,
+            self.logits,
+            scope='loss',
+            reduction=tf.losses.Reduction.SUM_OVER_BATCH_SIZE)
+        tf.summary.scalar('loss', loss)
+
+        self.global_step = tf.train.get_or_create_global_step()
+        optimizer = tf.train.AdamOptimizer(learning_rate=self.learning_rate)
```

Comment:
Have you tried running optimizer with BinaryCrossentropy() instead of Adam?

Reply:
DeepRec does some optimization on Adam optimizer, so we can see that DeepRec has a more obvious advantage compared to stock tf when using Adam optimizer. In Ali's model, they use adam optimizer.

Reply:
That's interesting. Could you point me to part of the code where I can see these optimizations?

Reply:
You can see the recently [Optimizer] commits in Alibaba/DeepRec. DeepRec does some fusion on this op.